src/app/loadout-builder/mod-assignments.ts

```diff
@@ -0,0 +1,140 @@
+  // Keep track of the current set of assignments extra energy investment so we dont
+  // need to continually recalculate it
+  private extraEnergyInvestment = DEFAULT_ENERGY_INVESTMENT;
+
+  constructor() {
+    this.slotDependantAssignments = new Map();
+    this.slotIndependantAssignments = new Map();
+  }
+
+  get isSuccessfullyAssigned() {
+    return this.extraEnergyInvestment < DEFAULT_ENERGY_INVESTMENT;
+  }
+
+  /**
+   * This returns a map of item id's to mod hashes for the assigned mods.
+   */
+  getResults() {
+    const rtn = new Map<string, number[]>();
+
+    for (const [item, itemAssignments] of this.slotIndependantAssignments.entries()) {
+      rtn.set(
+        item.id,
+        itemAssignments.mods.map((mod) => mod.hash)
+      );
+    }
+
+    for (const [itemId, mods] of this.slotDependantAssignments.entries()) {
+      if (rtn.has(itemId)) {
+        rtn.get(itemId)?.push(...mods);
+      } else {
+        rtn.set(itemId, mods);
+      }
+    }
+
+    return rtn;
+  }
+
+  /**
+   * Assigns slot dependant mods (i.e. helmet mods) to the item.
+   */
+  assignSlotDependantMods(item: DimItem, mods: PluggableInventoryItemDefinition[]) {
+    this.slotDependantAssignments.set(
+      item.id,
+      mods.map((mod) => mod.hash)
+    );
+  }
+
+  // Calculates the energy used and wasted for an item and its assignments.
+  private calculateEnergyChange(item: ProcessItem, assignments: ItemAssignments) {
+    const modCost =
+      (item.energy?.val || 0) + _.sumBy(assignments.mods, (mod) => mod.energy?.val || 0);
+    const energyUsedAndWasted = modCost + (item.energy?.originalCapacity || 0);
+    const energyInvested = Math.max(0, modCost - (item.energy?.originalCapacity || 0));
+    return assignments.energySwapped ? energyUsedAndWasted : energyInvested;
+  }
+
+  /**
+   * This method is used to keep track of the set of mods with the least amount of energy
+   * that needs to be invested to fit them.
+   *
+   * It does this by calculating the energy that is needed to be invested in the item and
+   * the energy wasted by swapping element. For example
+   * - If we have an item with 2 enegy and an aligned mod has 5 energy and a matching type
+   * we determine that to be a cost of 3.
+   * - If we have an item with 2 enegy and an aligned mod has 5 energy and a differing type
+   * we determine that to be a cost of 2 + 5 = 7.
+   *
+   * This will ensure we heavily favour matching energy types.
+   *
+   * TODO (ryan) I don't like that these are tied to process items. Lets make a new version
+   * of the mod assignment algorithm that takes dim items and house it in here.
+   */
+  assignSlotIndependantModsIfLessEnergyTypeSwaps(
+    newItems: ProcessItem[],
+    newGeneralMods: (ProcessMod | null)[],
+    newOtherMods: (ProcessMod | null)[],
+    newRaidMods: (ProcessMod | null)[]
+  ) {
+    const newAssignments = new Map<ProcessItem, ItemAssignments>();
+
+    for (const [i, item] of newItems.entries()) {
```

Comment:
I'm not sure I'm comfortable with this pattern for looping over an array. I get how and why it works but I don't love it.

Reply:
Part of it is it makes me think to hard about what's going on, part of it is that it creates a ton of garbage on the heap.

Reply:
Yeah I am happy to no do it then, as I need the index it just means I need the index version and pull the item out as the first thing.